### PR TITLE
Add notice REST endpoints and data store

### LINKED ISSUE #32226: Notices queue should have its own REST endpoints
Currently we preload the notices queue and use `updateOptions` to track notices that have been viewed.

https://github.com/woocommerce/woocommerce-admin/blob/c1f4ebfb7747584387be74f8a613e3726506b4d0/client/layout/transient-notices/index.js#L39-L41

https://github.com/woocommerce/woocommerce-admin/blob/c1f4ebfb7747584387be74f8a613e3726506b4d0/client/layout/transient-notices/index.js#L54-L56

This could benefit from creating 2 new endpoints and removing the use of `updateOptions` and `getOption` in this component.

* `/notices` - Get all queued notices (possibly with a filter param for user ID)
* `/notices/{$id}/dismiss` Dismiss a given notice.

### PR DESCRIPTION
**Pull request by joshuatf**
Fixes woocommerce/woocommerce#32226 

Adds a data store and REST endpoints for transient notices.

In hindsight this feels a bit odd given there is a data store for notices already, but I'm unaware of a way to preload that data store with data and fetching directly inside the component feels dirty.

### Screenshots
Screen Shot 2021-11-22 at 12 02 32 PM

### Detailed test instructions:

1.  Drop this into your test plugin:
```php
add_action( 'admin_init', 'add_test_transient_notice' );
function add_test_transient_notice() {
	TransientNotices::add(
		array(
			'user_id' => get_current_user_id(),
			'id'      => 'test-message',
			'status'  => 'success',
			'content' => 'The cows are taking over.  Moo.',
		)
	);
}
```
2. Navigate to any WCA page or embed page
3. Note the notice appears
4. Dismiss the notice and note the request to the `/dismiss` endpoint
5. If you refresh the page, the note will be re-added due to the above script, so you can either deactivate that test plugin to verify the notice has been dismissed prior to refresh or make a call to `wp-json/wc-admin/notices/` to check that the notices are empty

### CODE REVIEW

File: client/layout/transient-notices/index.js [hunk 1]
```diff
@@ -23,51 +22,51 @@ function TransientNotices( props ) {
...
 	useEffect( () => {
-		getCurrentUserNotices().forEach( ( queuedNotice ) => {
+		if ( ! hasFinishedResolution ) {
+			return;
+		}
+
+		noticesQueue.forEach( ( queuedNotice ) => {
 			const notice = applyFilters( QUEUED_NOTICE_FILTER, queuedNotice );
 
 			createNotice2( notice.status, notice.content, {
```
Comment (louwie17) [inline client/layout/transient-notices/index.js:62]: Could this logic be moved to the data store potentially? Given we could trigger an action from a resolver.

**Review by louwie17 — COMMENTED** [review]
Code looks good and tested well.
I am just wondering if it's possible to combine the two `notices` store, or maybe do away with `notices2`?
Mostly because we are creating the `notice2` inside the component and it be nice if we can move that to the store.

File: client/layout/transient-notices/index.js (hunk 1 above)
Comment (joshuatf) [reply client/layout/transient-notices/index.js:62]: That's an interesting idea, @louwie17.  I guess the question is whether or not we would ever want the ability to resolve/get the notices without first displaying them.

**Comment by joshuatf** [comment]
@becdetat I think you originally made the copy of `core/notices2`. Just curious if you can share any context and if the original issue may have been resolved.

I think there's room for clean-up in this PR to merge the two data stores (transients and notices2) or remove notices2 if it's no longer needed.

**Comment by becdetat** [comment]
`core/notices2` was created because I needed to modify `TransientNotices` to support some new features that needed to be released before the upstream `@wordpress/components` package was updated. The features were the icon and `onDismiss` callback. blame here: https://github.com/WordPress/gutenberg/blame/f7e057bd50502bdaf3ed6ed1c474ad1a3267ef42/packages/components/src/snackbar/index.js#L50.

Using the updated `@wordpress/components` was delayed because of the release cadence and support windows - we need to support I think n-2 of whatever `@wordpress/components` version is included in WordPress core. You'll need to double check if we're clear of the support windows before dropping `core/notices2`. I tried removing it [here](https://github.com/woocommerce/woocommerce-admin/pull/6756) a bit prematurely (back in April).

**Comment by mattsherman** [comment]
@joshuatf @louwie17 What's the status of this one? Is it still "In Progress"?

**Comment by joshuatf** [comment]
@mattsherman This is still in progress.  I had hoped to tackle it during this cooldown, but had to shift focus for the majority of this cooldown.

I will likely get to it by next cooldown unless we decide to include it in this cycle.

**Comment by mattsherman** [comment]
> @mattsherman This is still in progress. I had hoped to tackle it during this cooldown, but had to shift focus for the majority of this cooldown.
> 
> I will likely get to it by next cooldown unless we decide to include it in this cycle.

I'll put this at the bottom of Cycle Backlog.

**Comment by joshuatf** [comment]
After more thought on this, fetching of the transient notices via REST API is not a necessity.  Given that this data is necessary on every page, it is a candidate for preloading if we choose.

However, any actions we keep (dismissal should definitely be kept), should be merged with existing notice stores if possible.

**Comment by louwie17** [comment]
Update on the notices2 part, now the Wordpress 5.9 is released we are able to do away with the `notices2` as this was available in 5.7 (see [this list for Gutenberg](https://raw.githubusercontent.com/WordPress/gutenberg/trunk/docs/contributors/versions-in-wordpress.md) versions as reference and this comment by Timmy: https://github.com/woocommerce/woocommerce-admin/pull/6756#issuecomment-816184298

**Comment by louwie17** [comment]
Closing this as it has been moved to the WooCommerce repo -> https://github.com/woocommerce/woocommerce/pull/32321

**Pull request closed without merging**